data-service/src/main/java/com/amazonaws/blox/dataservice/api/StartDeploymentApi.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.dataservice.api;
+
+import com.amazonaws.blox.dataservice.mapper.ApiModelMapper;
+import com.amazonaws.blox.dataservice.model.Environment;
+import com.amazonaws.blox.dataservice.model.EnvironmentId;
+import com.amazonaws.blox.dataservice.model.EnvironmentRevision;
+import com.amazonaws.blox.dataservice.repository.EnvironmentRepository;
+import com.amazonaws.blox.dataservicemodel.v1.exception.InternalServiceException;
+import com.amazonaws.blox.dataservicemodel.v1.exception.ResourceNotFoundException;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentResponse;
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import java.util.UUID;
+
+@Slf4j
+@Component
+@AllArgsConstructor
+public class StartDeploymentApi {
+
+  @NonNull private final ApiModelMapper apiModelMapper;
+  @NonNull private final EnvironmentRepository environmentRepository;
+
+  public StartDeploymentResponse startDeployment(final StartDeploymentRequest request)
+      throws ResourceNotFoundException, InternalServiceException {
+    try {
+      final EnvironmentId environmentId =
+          apiModelMapper.toModelEnvironmentId(request.getEnvironmentId());
+      final Environment environment = environmentRepository.getEnvironment(environmentId);
+
+      final EnvironmentRevision environmentRevision =
+          environmentRepository.getEnvironmentRevision(
+              environmentId, request.getEnvironmentRevisionId());
+
+      environment.setActiveEnvironmentRevisionId(environmentRevision.getEnvironmentRevisionId());
+
+      environmentRepository.updateEnvironment(environment);
```

Comment:
This is predated your CR but I'm thinking about this: in environmentRepository.updateEnvironment(environment), dynamoDBMapper.load is called once to get the lastest recordVersion. This can truly get the optimistic lock but also add an extra transaction to DDB. Since we have to do a load in startDeploymentApi, which is the immediate upper level of environmentRepository.updateEnvironment, do you think we need load it again in environmentRepository.updateEnvironment?

Reply:
Good question.  Looked at current optimistic locking implementation in  `EnvironmentRepositoryDDB#updateEnvironment `,  I think it does not do what we want.

Considering following scenario:
 - Request A: start-deployment on environment E
 - Request B: Do a delete-environment at the same time
 - Request A gets E revision 1, passed validation (we will do it in the StartDeployment API ), environment looks good and ready to update environment
 - Request B get E revision 1, make it as deleted and update DDB. the E record revision is now 2
 - **ISSUE**: in current implementation, say `environmentRepository.updateEnvironment` fetch the E with revision 2, it will make a success call to update the DDB record.

To make optimistic locking actually does it job, we need to make sure the data used for request validation is the one we finally used for update.

I think we should have recordRevision in the Environment model and it will map to the environment record for optimistic checking.

Reply:
Since it's a different issue of this PR, should we have a different PR for the optimistic locking issue?

Reply:
Agree, so I think remove the setRecordVersion piece of code from environmentDepositoryDDB and add recordRevision in environment model mapper will 1: fix the locking issue, 2: avoid extra read from DDB.